ddm_test_validate.py

```diff
@@ -0,0 +1,214 @@
+import pathlib
+import os
+import time
+from distutils.util import strtobool
+from typing import Any, Dict, List
+
+import numpy as np
+import pandas as pd
+from azure.core.exceptions import HttpResponseError
+from dotenv import load_dotenv, set_key
+from microsoft_bonsai_api.simulator.client import BonsaiClient, BonsaiClientConfig
+from microsoft_bonsai_api.simulator.generated.models import (
+    SimulatorInterface,
+    SimulatorSessionResponse,
+    SimulatorState,
+)
+
+from base import BaseModel
+
+logging.basicConfig()
+logging.root.setLevel(logging.INFO)
+logger = logging.getLogger("datamodeler")
+
+import hydra
+from omegaconf import DictConfig
+
+from model_loader import available_models
+
+from sim.moab.moab_main import SimulatorSession,  env_setup
+from sim.moab.policies import coast, random_policy
+
+dir_path = os.path.dirname(os.path.realpath(__file__))
+env_name = "DDM"
+log_path = "logs"
+
+
+class Simulator(BaseModel):
+    def __init__(self, model, states=List[str], actions=List[str], configs=List[str], 
+        log_file: str = None,):
+        self.dd_model = model
+        self.features = states + actions + configs
+        self.labels = states
+        self.config_keys = configs
+        self.state_keys = states
+        self.action_keys = actions
+        self.sim_physics = SimulatorSession()
+        if log_file=="enable":
+            current_time = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
+            log_file = current_time + "_" + env_name + "_log.csv"
+            log_file2 = current_time + "_" + "SIM" + "_log.csv"
+            logs_directory = pathlib.Path(log_file).parent.absolute()
+            if not pathlib.Path(logs_directory).exists():
+                print(
+                    "Directory does not exist at {0}, creating now...".format(
+                        str(logs_directory)
+                    )
+                )
+                logs_directory.mkdir(parents=True, exist_ok=True)
+        self.log_file = os.path.join(log_path, log_file)
+        self.log_file2 = os.path.join(log_path, log_file2)
+
+    def episode_start(self, config: Dict[str, Any] = None):
+        if config:
+            self.config = config
+        else:
+            # configs randomized here. Need to decide a single place for configs
+            # range either in main.py or in simulator configs
+            self.config = {j:np.random.uniform(-1,1) for j in self.config_keys}
+        # Assign same state as would be used by simulator
+        self.sim_physics.episode_start(config)
+        _fullstate = self.sim_physics.get_state()
+        # idempotent dict comprehension for sims with prefix in configurations
+        self.state = {j:_fullstate[l] for j in self.state_keys for l in _fullstate.keys() if l in j}
+        
+
+    def episode_step(self, action: Dict[str, int]):
+
+        input_list = [
+            list(self.state.values()),
+            list(self.config.values()),
+            list(action.values()),
```

Comment:
with numpy arrays it's much safter to do `np.ndarray.values().tolist()` rather than `list(np.ndarray.values())`. The former guarantees you get a list (potentially nested) of primitive python types back. the latter might give you `List[np.ndarray, np.ndarray]`  if you have more than 1 dimension for example, which isn't the case here but just in case...

Reply:
That is a good idea, and I will use in other places. In this particular case, the states, config and action are `dict_values`, due to the structure of rest of the ddm implementation.

Reply:
The state ,config and action are dict_values, so `.tolist` isn't available here. But, that is a good point, and I will use that in other places.